### ISSUE TITLE
Aggregate metric datapoints over time period

### ISSUE THREAD

**Issue by 0x006EA1E5** [open-telemetry/opentelemetry-collector-contrib#29461]
### The purpose and use-cases of the new component

This processor would receive a stream of data points for a a metric timeseries, and periodically emit an "aggregate" at a set interval.

We can achieve something similar already by exporting metrics to the Prometheus exporter, then periodically scraping the prom endpoint with a Prometheus receiver. However this is clunky and somewhat less efficient than using a processor.

One concrete use case is where we want to send high frequency metric datapoints to the Prometheus remote write exporter, for example datapoints produces by the count connector. When counting spans, the count connector will produce a single delta datapoints (increment value 1) for each counted span, which could of course be many times per second.  However, typically we would only want to remote write to Prometheus periodically, as we would if we were scraping, perhaps once every 30 seconds. This is especially true for downstream metric sinks that charge for datapoints per minute.

This proposed processor would be stateful, tracking metrics by identity, maintaining a single aggregate value. This aggregate will be output every interval.

For example, if the processor received the following delta datapoints: 1, 3, 5, then at the next "tick" of the interval clock, a single delta datapoint of 9 would be emitted.

Similarly, if the processor received the following cumulative datapoints: 7, 9, 11, then at the next "tick" of the interval clock, a single cumulative datapoint of 11 would be emitted.

There would be a "max_staleness" config option so that we can stop tracking metrics which don't receive any data for a given time.

### Example configuration for the component

max_staleness
Include/exclude MatchMetrics

### Telemetry data types supported

Metrics

### Is this a vendor-specific component?

- [ ] This is a vendor-specific component
- [ ] If this is a vendor-specific component, I am proposing to contribute and support it as a representative of the vendor.

**Comment by atoulme** [comment]
@crobert-1 who is the sponsor for this issue?

**Comment by crobert-1** [comment]
> @crobert-1 who is the sponsor for this issue?

@djaglowski volunteereed to be the sponsor [here.](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29300#issuecomment-1816624955)

**Comment by verejoel** [comment]
This is something that is on our radar, and would like to support as much as possible. Our use case is to enable remote writing of metrics from the count connector to Thanos.

**Comment by RichieSams** [comment]
Some clarifications of the spec for this one, given that https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30479 will exist:

- All non-cumulative metrics will be dropped. If you have delta metrics, use https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30479 to convert them to cumulative
- The aggregator will only store the "last" value received per unique combo of metric name + labels

@djaglowski Do you have any thoughts for a name? `timeaggregationprocessor`?

**Comment by djaglowski** [comment]
@RichieSams, this all looks good to me.

**Comment by sh0rez** [comment]
other name idea: intervalprocessor (because it emits at fixed intervals)

for implementing this, you may want to take a look at `streams.Ident` from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30707, which can be used as a `map[streams.Ident]T` for storing last values for each distinct sample stream

**Comment by RichieSams** [comment]
Thanks for the pointer! I like that name as well; it's less of a mouthful.

**Comment by verejoel** [comment]
Just so I understand the current situation:
* #29300 dropped in favour of #30479 
* #30479 will convert delta metrics to cumulative _and_ perform aggregation on a configurable interval
* #29461 (this processor) will just do the aggregation, and could (contrived example) be used to re-aggregate the metrics coming out of #30479

Is this correct?

**Comment by RichieSams** [comment]
Not quite, https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30479 will only convert delta to cumulative. IE in ideal settings, there will be a 1 to 1 mapping of input metrics to output metrics.

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29461 (this processor) will only do aggregation.

**Comment by RichieSams** [comment]
As an added comment: I have started working on this issue, PRs to follow in the coming days

**Comment by djaglowski** [comment]
Thanks @RichieSams!

**Comment by 0x006EA1E5** [comment]
> - All non-cumulative metrics will be dropped. If you have delta metrics, use https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30479 to convert them to cumulative

There is no reason why we can't aggregate deltas over time too, though, right?

**Comment by RichieSams** [comment]
> > * All non-cumulative metrics will be dropped. If you have delta metrics, use [new component: deltatocumulative processor #30479](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30479) to convert them to cumulative
> 
> There is no reason why we can't aggregate deltas over time too, though, right?

It just duplicates the code of the new deltatocumulativeprocessor. Unless you mean something else like:

1. Accumulate delta metrics
2. At each interval export the current sum for each delta
3. Reset all the deltas back to zero

This *could* work. But I'd be curious to the use-cases for that. Vs just converting to cumulative and aggregating those.

**Comment by 0x006EA1E5** [comment]
> > > * All non-cumulative metrics will be dropped. If you have delta metrics, use [new component: deltatocumulative processor #30479](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30479) to convert them to cumulative
> > 
> > There is no reason why we can't aggregate deltas over time too, though, right?
> 
> It just duplicates the code of the new deltatocumulativeprocessor. Unless you mean something else like:
> 
> 1. Accumulate delta metrics
> 2. At each interval export the current sum for each delta
> 3. Reset all the deltas back to zero
> 
> This *could* work. But I'd be curious to the use-cases for that. Vs just converting to cumulative and aggregating those.

Yes exactly, many deltas in, one delta out. I _think_ the batch processor does something similar, although the output trigger is batch size, not a clock period.

I guess one use case would be where deltas are preferred/required downstream, but you want to reduce the data rate, e.g., sending a single delta of 10,000 over the wire after 15s is much more efficient than sending 10,000 deltas of 1.

I'm not saying this has to be in scope for the initial implementation if there is no immediate need, but just trying to think about how all this might work.

**Comment by 0x006EA1E5** [comment]
> - All non-cumulative metrics will be dropped.

I'm assuming that the config across all these processes will be somewhat consistent.

The Cumulative to Delta Processor can be configured with metric include and exclude rules.

Isn't it more appropriate to simply pass through metric data that isn't matched?

If a user wants to drop a certain metric, they should configure a filter processor, no?

This gives maximum flexibility, and the pipeline is nice and explicit.

We'll, I guess it's best to align with what other processors do in comparable situations...

**Comment by RichieSams** [comment]
> I think the batch processor does something similar, although the output trigger is batch size, not a clock period.

It looks like the batchprocessor doesn't do any aggregation. It collects groups of metrics, and then sends them all at once in a single go. https://github.com/open-telemetry/opentelemetry-collector/blob/main/processor/batchprocessor/batch_processor.go#L425

IE, it would collect 10,000 deltas, and send them all at once to the next component in the pipeline. Vs without it, the next component would get 10,000 individual requests with each single metric. So it's optimizing for things like TCP connection / request overhead.

> I'm not saying this has to be in scope for the initial implementation if there is no immediate need, but just trying to think about how all this might work.

For sure. I think it could be added in a future scope if the need arises. I don't think that would impact the immediate design.

**Comment by RichieSams** [comment]
> Isn't it more appropriate to simply pass through metric data that isn't matched?

I don't have a strong opinion either way. I'd go with whatever the convention is for other processors.

So for delta metrics, we'd "consume" all the metrics and then periodically export on the interval. For cumulative metrics, we'd pass them to the next component in the chain untouched. Yes?

**Comment by 0x006EA1E5** [comment]
> > Isn't it more appropriate to simply pass through metric data that isn't matched?
> 
> I don't have a strong opinion either way. I'd go with whatever the convention is for other processors.
> 
> So for delta metrics, we'd "consume" all the metrics and then periodically export on the interval. For cumulative metrics, we'd pass them to the next component in the chain untouched. Yes?

I think it's the other way round😅.

The principle use-case is for cumulative metrics, as produced by the new delta to cumulative processor.

But otherwise, yes, exactly 👍

**Comment by RichieSams** [comment]
Right right, my brain is fried today. lol

**Comment by RichieSams** [comment]
@djaglowski @sh0rez @0x006EA1E5 How should this processor export the aggregated metrics?

The current implementation will export the metrics every X seconds to infinity. `max_staleness` affects if / when metrics will no longer be exported. IE:

1. As metrics come in, they are aggregated
2. New metrics are added to the state
3. Existing metrics replace whatever was already in the state (IFF their timestamp is newer)

1. In parallel to this, an exporter runs every X seconds
2. It exports all metrics currently stored in the state (Should it update all their timestamps to "now()"?)

However, I realized there is another potential approach. The processor could aggregate metrics over time, and then on an interval, export the aggregate once, flushing the state to empty.

1. As metrics come in, they are aggregated
2. New metrics are added to the state
3. Existing metrics replace whatever was already in the state (IFF their timestamp is newer)

1. In parallel to this, an exporter runs every X seconds
2. It exports all metrics currently stored in the state (Should it update all their timestamps to "now()"?)
3. The state is cleared to empty

In both case, a mutex is used to ensure aggregation and export are serialized with each other.

**Comment by 0x006EA1E5** [comment]
> However, I realized there is another potential approach. The processor could aggregate metrics over time, and then on an interval, export the aggregate once, flushing the state to empty.

I think the first approach it correct, the exporter publishes the latest / current value every interval.  


> It exports all metrics currently stored in the state (Should it update all their timestamps to "now()"?)

Yes, the timestamp should be updated to `now()`.  Or at least, the be precise, the exported datapoint should have the timestamp of the instant it was exported.  Looking at [the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#sums), cumulative datapoints can also have a `start_time` timestamp, which in our case would be the timestamp of the previous publish (or, if this is the first publish for the timeseries,  then I think the spec describes what should happen in the various edge cases).  The output is then a series of contiguous, non-overlapping intervals with no gaps.

This is then similar to how things work if we were to publish a metric to a prometheus exporter, and the scrape that back in with a prometheus receiver.  We can export datapoints to a prometheus exporter at any rate, but as the prometheus receiver scrapes at a set, steady interval, it just produces one datapoint per timeseries every scrap interval.  This includes the case where no new datapoints were sent to the prometheus exporter, the next scrape will just output a new datapoint with the `now()` timestamp (as I understand it).

Doing it this way meets a key use-case of working with native OTLP metrics in the whole pipeline, until eventually sending to a prometheusremotewrite exporter.

**Comment by 0x006EA1E5** [comment]
> The processor could aggregate metrics over time, and then on an interval, export the aggregate once, flushing the state to empty.

I think if/when we have the ability to work with delta datapoints, we would do something similar, or at least set the _value_ to zero (we would still keep state for the timeseries).  

But as I understand it, we are just looking at cumulative datapoints for now, and will pass-through deltas, right?

**Comment by RichieSams** [comment]
> cumulative datapoints can also have a start_time timestamp, which in our case would be the timestamp of the previous publish

Wouldn't `start_timestamp` just be inherited? IE if the incoming metrics have it, we leave it as-is

**Comment by 0x006EA1E5** [comment]
Looking at the spec, for cumulative datapoints the `start_time_unix_nano` should be the start of series, so the same value for each subsequent datapoint

> Contrast with cumulative aggregation temporality where we expect to report the full sum since 'start' (where usually start means a process/application start):

(I will edit my comment above to correct it)

So I suppose in our case, that is going to mean the cumulative datapoints' `start_time_unix_nano` will be either
* The `start_time_unix_nano` of the first delta datapoint received, if present
* Or, the `time_unix_nano` of the first delta datapoint received, if `start_time_unix_nano` is not present

**Comment by 0x006EA1E5** [comment]
It is not clear to me, but it looks like a Sum consists of NumberDataPoints, and `start_time_unix_nano` is optional to for those https://github.com/open-telemetry/opentelemetry-proto/blob/v0.9.0/opentelemetry/proto/metrics/v1/metrics.proto#L395

**Comment by RichieSams** [comment]
@sh0rez Can I get your opinion on the two export options I presented above? We have one vote for each atm :P

**Comment by 0x006EA1E5** [comment]
> time is important to keep, as we shall not "lie values" into the future.

Yes this makes sense 👍 

@sh0rez 
> imo the reason for this processor to exist is to limit the flowrate of datapoints (datapoints per minute, etc).

I would say the reason if to _fix_ the flowrate, rather than just _limit_.

I think we have an open question as to whether we should publish a new cumulative datapoint at every interval in the case where the cumulative value has not changed.  e.g., would we publish a stream of `0`s every interval, or just publish the first `0` and the only publish another datapoint when there is a _new_ / different value.

I detailed my use case in more detail here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30827#discussion_r1508656086
